components/raftstore/src/store/peer.rs

```diff
@@ -2336,11 +2338,32 @@ where
             }
             Ok(RequestPolicy::ReadIndex) => return self.read_index(ctx, req, err_resp, cb),
             Ok(RequestPolicy::ProposeNormal) => {
-                let store_id = ctx.store_id();
-                if disk::disk_full_precheck(store_id) || ctx.is_disk_full {
-                    Err(Error::Timeout("disk full".to_owned()))
-                } else {
+                let mut allowed = true;
+                let mut err_msg = String::from("propose failed: disk ");
+                let store_id = ctx.store.get_id();
+                match disk_full_opt {
+                    DiskFullOpt::AllowedOnAlreadyFull => allowed = true,
+                    DiskFullOpt::AllowedOnAlmostFull => {
+                        if disk::is_disk_already_full(ctx.disk_status, store_id) {
+                            allowed = false;
+                            err_msg.push_str("already full");
+                        }
+                    }
+                    DiskFullOpt::NotAllowedOnFull => {
+                        if disk::is_disk_already_full(ctx.disk_status, store_id) {
```

Comment:
It's better to use `fn disk::usage_status() -> UsageStatus`.

Reply:
Better to let fail_point only appear in disk.rs?

Reply:
We can move failpoints into `get_disk_status`.